riscv/cfg.h

```diff
@@ -61,15 +65,34 @@ class cfg_t
       nprocs(default_nprocs),
       isa(default_isa),
       priv(default_priv),
-      mem_layout(default_mem_layout)
+      mem_layout(default_mem_layout),
+      frozen(false)
   {}
 
+  ~cfg_t();
+
   cfg_arg_t<std::pair<reg_t, reg_t>> initrd_bounds;
   cfg_arg_t<const char *>            bootargs;
   cfg_arg_t<size_t>                  nprocs;
   cfg_arg_t<const char *>            isa;
   cfg_arg_t<const char *>            priv;
   cfg_arg_t<std::vector<mem_cfg_t>>  mem_layout;
+
+  // A function to fill out the private members that are derived from the
+  // cfg_arg_t<...> members above.
+  //
+  // Call this once after filling those out and then don't touch the
+  // configuration again.
```

Comment:
"and then don't touch the configuration again" -- this seems ripe for misuse.  Would it not be better to have `freeze` return a copy, so that there's no possibility the frozen thing could change underfoot?

I'm not worried about the extra cost of the copy in this context.

Reply:
Hmm, maybe. I'd convinced myself that, since there was exactly one place that uses this, we could be reasonably sure things would work. But there's always scope for messing up! :-)

Alternatively, we could do what I alluded to in the commit message and make an "expanded config object" (so there's no frozen/unfrozen state to track). The awkward bit is that you either have to copy across all the fields (which adds a certain amount of boilerplate to the constructor) or you end up with things like `cfg.base_cfg.nprocs()` all over the place, which is horrible.

I'm slightly coming down on the idea of an expanded config object class (probably called `cfg_t`, and we'll call the thing we had `pre_cfg_t` or something). That can have const fields, meaning it's impossible to mess stuff up. And I think we should copy over the fields and just take the boilerplate hit :-) Does that sound reasonable to you? (I'd like to check before spending half an hour doing boring edits!)